File: sdk/go/message.go
package conduit

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// MessageType represents the type of message
type MessageType string

const (
	// MessageTypeData is a regular data message
	MessageTypeData MessageType = "data"

	// MessageTypeControl is a control plane message
	MessageTypeControl MessageType = "control"

	// MessageTypeError is an error message
	MessageTypeError MessageType = "error"
)

// Message represents a message in the Exchange
type Message struct {
	// ID is a unique identifier for this message
	ID string `json:"id"`

	// Timestamp is when the message was created
	Timestamp time.Time `json:"timestamp"`

	// Sequence is the NATS stream sequence number
	Sequence uint64 `json:"sequence"`

	// Type is the message type
	Type MessageType `json:"type"`

	// Payload is the message content (arbitrary JSON)
	Payload json.RawMessage `json:"payload"`

	// Metadata contains optional metadata
	Metadata map[string]string `json:"metadata,omitempty"`
}

// UnmarshalPayload unmarshals the payload into the given struct
func (m *Message) UnmarshalPayload(v interface{}) error {
	return json.Unmarshal(m.Payload, v)
}

// ControlMessage represents a control plane message
type ControlMessage struct {
	Command string                 `json:"command"`
	Args    map[string]interface{} `json:"args,omitempty"`
}

const (
	// ControlCommandPause pauses message processing
	ControlCommandPause = "pause"

	// ControlCommandResume resumes message processing
	ControlCommandResume = "resume"

	// ControlCommandShutdown requests graceful shutdown
	ControlCommandShutdown = "shutdown"
)

// NewMessage creates a new message with the given payload
func NewMessage(msgType MessageType, payload interface{}) (*Message, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	return &Message{
		ID:        generateMessageID(),
		Timestamp: time.Now(),
		Type:      msgType,
		Payload:   payloadBytes,
		Metadata:  make(map[string]string),
	}, nil
}

// MarshalMessage marshals a message to JSON bytes
func MarshalMessage(msg *Message) ([]byte, error) {
	return json.Marshal(msg)
}

// UnmarshalMessage unmarshals a message from JSON bytes
func UnmarshalMessage(data []byte) (*Message, error) {
	var msg Message
	if err := json.Unmarshal(data, &msg); err != nil {
		return nil, err
	}

	// Validate the message
	if err := ValidateMessage(&msg); err != nil {
		return nil, err
	}

	return &msg, nil
}

// ValidateMessage validates that a message has all required fields
func ValidateMessage(msg *Message) error {
	if msg.ID == "" {
		return fmt.Errorf("message ID is required")
	}
	if msg.Type == "" {
		return fmt.Errorf("message type is required")
	}
	if msg.Type != MessageTypeData && msg.Type != MessageTypeControl && msg.Type != MessageTypeError {
		return fmt.Errorf("invalid message type: %s", msg.Type)
	}
	if len(msg.Payload) == 0 {
		return fmt.Errorf("message payload is required")
	}
	return nil
}

// generateMessageID generates a unique message ID
func generateMessageID() string {
	return uuid.New().String()
}

File: sdk/typescript/src/message.ts
/**
 * Message envelope types for Conduit SDK.
 *
 * Defines the standard message format used for communication.
 */

import { v4 as uuidv4 } from 'uuid';

/**
 * Type of message.
 */
export enum MessageType {
  DATA = 'data',
  CONTROL = 'control',
  ERROR = 'error',
}

/**
 * Message envelope for Conduit communication.
 */
export interface Message<T = any> {
  /** Unique message identifier */
  id: string;
  /** When the message was created (ISO 8601 format) */
  timestamp: string;
  /** NATS sequence number */
  sequence: number;
  /** Message type (data, control, or error) */
  type: MessageType;
  /** The actual message payload */
  payload: T;
  /** Optional metadata */
  metadata?: Record<string, string>;
}

/**
 * Create a new message.
 *
 * @param type Message type
 * @param payload Message payload (must be JSON-serializable)
 * @param metadata Optional metadata
 * @returns New message instance
 */
export function createMessage<T>(
  type: MessageType,
  payload: T,
  metadata?: Record<string, string>
): Message<T> {
  return {
    id: uuidv4(),
    timestamp: new Date().toISOString(),
    sequence: 0, // Will be set by NATS
    type,
    payload,
    metadata,
  };
}

/**
 * Parse a message from JSON bytes.
 *
 * @param data JSON-encoded message
 * @returns Parsed message instance
 * @throws Error if the message cannot be parsed
 */
export function parseMessage<T = any>(data: Uint8Array | string): Message<T> {
  try {
    const text = typeof data === 'string' ? data : new TextDecoder().decode(data);
    const parsed = JSON.parse(text);

    if (!parsed.id || !parsed.timestamp || !parsed.type) {
      throw new Error('Missing required message fields');
    }

    return {
      id: parsed.id,
      timestamp: parsed.timestamp,
      sequence: parsed.sequence || 0,
      type: parsed.type as MessageType,
      payload: parsed.payload,
      metadata: parsed.metadata,
    };
  } catch (error) {
    throw new Error(`Failed to parse message: ${error}`);
  }
}

/**
 * Serialize message to JSON bytes.
 *
 * @param msg Message to serialize
 * @returns JSON-encoded message
 */
export function serializeMessage<T>(msg: Message<T>): Uint8Array {
  const data: any = {
    id: msg.id,
    timestamp: msg.timestamp,
    sequence: msg.sequence,
    type: msg.type,
    payload: msg.payload,
  };

  if (msg.metadata) {
    data.metadata = msg.metadata;
  }

  return new TextEncoder().encode(JSON.stringify(data));
}
